### ISSUE TITLE
Introduce additional HierarchyModifier::move1LevelUp etc.

### ISSUE THREAD

**Issue by mhier** [ChimeraTK/ApplicationCore#22]
Introduce additional HierarchyModifiers:
- move1LevelUp
- move2LevelUp
- ... up to 10 is probably sufficient.

**Comment by killenb** [comment]
I don't know about the implementation specifics, but could we template this:

```
template<int deltaLevel>
moveLevelsUp<deltaLevel>

```

**Comment by mhier** [comment]
HierarchyModifier is a enum, so this won't work as far as I know.

**Comment by killenb** [comment]
I was just implementing oneLevelUp.  When writing tests I noticed that instead of
```
movedVarGroup{this, "moved", "desc", oneLevelUp}
```
I could just write
```
movedVarGroup{owner, "moved", "desc", hideThis}
```
The tests work like this.
Is there anything wrong with the approach? Will it give problems when shutting down?

**Comment by killenb** [comment]
Or is it giving problems because it is mixing the virtual hierarchy with the C++ hierarchy?

**Comment by killenb** [comment]
Ok, you don't even need the hack with creating the "movedVarGroup". You can directly put "owner" to the constructor of the variables inside the movedVarGroup.

**Comment by mhier** [comment]
It's simply not the same thing. Imaging "this" is an ApplicationModule, getOwner() is then either the application or an ModuleGroup. VariableGroups cannot be held by ModuleGroups or the application directly, hence this approach wouldn't work there. It would only allow you to move VariableGroups around within your ApplicationModule - in that case you could also make it right from the beginning.

**Comment by killenb** [comment]
I don't understand what you mean with this:
>  in that case you could also make it right from the beginning.

But I see that this is exactly the clash between the C++ and the virtual hierarchy which I mentioned in my comment. So with the current ApplicationCore it will not work in all cases (I already noticed that I cannot put entityOwners as owner to variables).

What I don't understand is what the "this" / owner stuff is doing exactly. Could we change it so you can build virtual hierarchies with it? And the pointer type would be EntityOwner. Because you are getting a pointer to EntityOwner (virtual plane), which is pointing to a C++ object. There might be a  quirk in the design logic.

**Comment by mhier** [comment]
The ownership needs to follow the C++ ownership. Otherwise you might get undefined behaviour. E. g. readAll() will derive what to read from that ownership. If you would move a VariableGroup into another ApplicationModule by changing its ownership pointers, the other module might accidentally read/write these variables. That would be a violation of thread safety. You should never do this, even if you think you got it right. 

What I meant with that sentence: you do not need hierarchy modifiers to achieve a certain variable hierarchy *within* an ApplicationModule. You could just build the hierarchy directly as intended. Maybe this is not 100% correct, as it can be convenient to introduce groups e.g. For readAll() etc., but this should be limited to the hideThis modifier.

**Comment by killenb** [comment]
OK, I see that the pointer MUST always be `this`. Otherwise we get into serious C++ trouble. I still see a small inconsistency: findTag gives you a list of EntityOwners as virtual hierarchy, but entity owners are C++ hierarchy.
If I don't have this you can either use tags, or have to cable the input variable manually. In both cases you have that thing in the CS twice (once where it should be, and once as input for the monitor).

Seems I will have to implement the oneLevelUp modifier and cannot get away with using getOwner().

I have the case that I want to put a pre-build ApplicationModule (StatusMonitor) into the same virtual folder as its input variable  which I have in my application. And it then should auto-connect to that variable. For this I need the oneLevelUp, to get the input variable out of the hierarchy of the monitor. I cannot get this with a-priori design because the StatusMonitor is a generic, pre-build module.

**Comment by mhier** [comment]
findTag gives you VirtualModules, those are distinct objects from the original VariableGroups etc.. The VirtualModules have always other VirtualModules as owner. This is not an inconsistency.

**Comment by killenb** [comment]
Ah, ok, this explains my confusion. I thought EntityOwner was the virtual note. Then the two hierarchies are cleanly separated. Then we definitely need the modifier.

**Comment by killenb** [comment]
I need the owner of a virtual module, but they are all nullptr. I was trying to set them and found out that this will blow up if you use the copy constructor and assignment operators of VirtualModules, because then the virtual ownership and C++ ownership are not the same and pointers will go out of scope.

Is there a reason to have the copy constructor and assignment operator? Or should I not use the owner mechanism for other reasons? It kind of mixes the two planes, but then a VirtualModule is an EntityOwner, so they are already mixed. Should the virtual module not be separate from it?

**Comment by killenb** [comment]
Do we need oneLevelUp with and without hiding the moved hierarchy?
I managed to move one level up without hiding, which is not what I wanted to do. But I see that it could be useful.

**Comment by killenb** [comment]
For a proper implementation with nested oneUpAndHide I need an owner stored in the class, so I can navigate backwards in the tree. This is not safely implementable with the existence of copy constructor and assignment operator.

How do we proceed:
* Remove the copy constructor and assignment operator?
* Only implement non-nested oneUpAndHide?
* Have nullptr in all copied and assigned objects (owner functionality broken, but no invalid pointers)?

**Comment by killenb** [comment]
The current implementation does not create a virtual owner hierarchy and oneUpAndHide cannot be nested. I created a follow up ticket #72 to fix this.
Apart from that the functionality is working.

Closed in 49390fac0c35
